In [2]:
import pandas as pd
import numpy as np
from ast import literal_eval

In [53]:
df = pd.read_csv('wondoofin.csv')
cat_df = pd.read_csv('category.csv')

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 559 non-null    int64 
 1   이름                 559 non-null    object
 2   로스터리               559 non-null    object
 3   타입                 559 non-null    object
 4   로스팅 포인트            559 non-null    object
 5   지속가능성              113 non-null    object
 6   컵 노트               559 non-null    object
 7   바디감                559 non-null    int64 
 8   신맛                 559 non-null    int64 
 9   단맛                 559 non-null    int64 
 10  쓴맛                 559 non-null    int64 
 11  커피 소개              559 non-null    object
 12  나라                 543 non-null    object
 13  로스터리 소개            559 non-null    object
 14  로스터리 주소            550 non-null    object
 15  이름.1               559 non-null    object
 16  식품의 유형             554 non-null    object
 1

In [55]:
df.columns

Index(['id', '이름', '로스터리', '타입', '로스팅 포인트', '지속가능성', '컵 노트', '바디감', '신맛', '단맛',
       '쓴맛', '커피 소개', '나라', '로스터리 소개', '로스터리 주소', '이름.1', '식품의 유형',
       '제조원 및 소재지', '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량',
       '제품문의 관련 주소 및 전화번호', '카페인', '로스터리ID'],
      dtype='object')

In [56]:
df.shape

(559, 26)

***
null 값 처리
- 지속가능성 : 113
- 나라 : 543
***
- 로스터리 주소 : 550
- 제조원 및 소재지 : 557/559
- 유통기한 : 557
- 보관 방법 : 545
- 식품의 유형 : 554
***

In [57]:
# 리스트 처리
df['컵 노트'] = df['컵 노트'].apply(literal_eval)
df['컵 노트'][0]

['토피', '맥아', '흑설탕']

***

In [58]:
#################################
########## 지속 가능성 ##########
#################################
for i in range(len(df)):
    if df['지속가능성'][i] is np.nan:
        df.at[i, '지속가능성'] = 0
        
df['지속가능성']

0         0
1         0
2       유기농
3      공정무역
4         0
       ... 
554       0
555       0
556       0
557       0
558       0
Name: 지속가능성, Length: 559, dtype: object

In [ ]:
df['나라'] = df['나라'].str.split(', ')
df['나라'][0]

In [ ]:

for i in range(len(df)):
    if df['나라'][i] is np.nan:
        df.at[i, '나라'] = 0
        
df['나라']

***
추천과는 상관없는 칼럼들
1. 주소
***

In [61]:
print(df.loc[df['로스터리 주소'].isna(), 'id'].tolist())

[4701, 4692, 4613, 4614, 4695, 4694, 4612, 4611, 4693]


In [62]:
print(df.loc[df['제조원 및 소재지'].isna(), 'id'].tolist())

[274, 272]


In [63]:
print('로스터리 : ', df.loc[df['id'] == 4701, '로스터리'].values[0])
print('로스토리 주소 : ', df.loc[df['id'] == 4701, '로스터리 주소'].values[0])
print('제조원 : ', df.loc[df['id'] == 4701, '제조원 및 소재지'].values[0])

로스터리 :  커피네츄럴스페셜티
로스토리 주소 :  nan
제조원 :  경기도 고양시 덕양구 행신동 741-17 1층


***
df의 로스터리 null 정보 확인
***

In [64]:
# df.loc[df['로스터리 주소'].isna(), '로스터리 주소'] = 0
# df.loc[df['제조원 및 소재지'].isna(), '제조원 및 소재지'] = 0
df[['로스터리 주소', '제조원 및 소재지']] = df[['로스터리 주소', '제조원 및 소재지']].fillna(0)

In [65]:
addrdf = df[['로스터리ID', '로스터리', '로스터리 주소', '제조원 및 소재지']]
addrdf.head()

로스터리ID         로스터리                                         로스터리 주소  \
0      57        언더프레셔                           서울 강남구 영동대로106길 37 1층   
1      67         운조커피                          경기도 화성시 동탄지성로 306번길 68   
2      77    제이엠디커피팩토리  경기 고양시 일산동구 장대길 75-53 나동 1층 JMD Coffee Factory   
3      11  뉴웨이브 커피로스터스                           서울특별시 양천구 오목로56길 2,1층   
4      34     베르크 로스터스                          부산 부산진구 전포동 356-6 지하1층   

                                        제조원 및 소재지  
0                    (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  
1                          경기도 화성시 동탄지성로 306번길 68  
2  경기 고양시 일산동구 장대길 75-53 나동 1층 JMD Coffee Factory  
3        뉴웨이브커피로스터스, 서울시 양천구 오목로 56길 2, 영원빌딩 103호  
4             베르크로스터스, 부산광역시 사상구 학감대로222번길 83, 2층

In [66]:
result = addrdf[addrdf['로스터리 주소']==0]
result

로스터리ID       로스터리 로스터리 주소                  제조원 및 소재지
17       89  커피네츄럴스페셜티       0  경기도 고양시 덕양구 행신동 741-17 1층
59       89  커피네츄럴스페셜티       0  경기도 고양시 덕양구 행신동 741-17 1층
137     110       헤베커피       0       서울 중구 필동로 32 낙원빌딩 1층
240     110       헤베커피       0       서울 중구 필동로 32 낙원빌딩 1층
248      89  커피네츄럴스페셜티       0  경기도 고양시 덕양구 행신동 741-17 1층
322      89  커피네츄럴스페셜티       0  경기도 고양시 덕양구 행신동 741-17 1층
455     110       헤베커피       0       서울 중구 필동로 32 낙원빌딩 1층
521     110       헤베커피       0       서울 중구 필동로 32 낙원빌딩 1층
527      89  커피네츄럴스페셜티       0   경기도 고양시 덕양구 행신동741-17 1층

In [67]:
# result2 = addrdf[addrdf['제조원 및 소재지'].isna()]
result2 = addrdf[addrdf['제조원 및 소재지']==0]
result2

로스터리ID   로스터리                 로스터리 주소 제조원 및 소재지
413      42  센터 커피  서울특별시 성동구 서울숲 2길 28-11         0
529      42  센터 커피  서울특별시 성동구 서울숲 2길 28-11         0

In [68]:
# 항상 '로스터리 주소'와 '제조원 및 소재지' 주소가 일치하는가? 
result3 = addrdf[addrdf['제조원 및 소재지'] != addrdf['로스터리 주소']]
result3

로스터리ID         로스터리                    로스터리 주소  \
0        57        언더프레셔      서울 강남구 영동대로106길 37 1층   
3        11  뉴웨이브 커피로스터스      서울특별시 양천구 오목로56길 2,1층   
4        34     베르크 로스터스     부산 부산진구 전포동 356-6 지하1층   
6        18    라메르판지 비앤알              울산 북구 판지1길 30   
7        98         토치커피        서울시 강남구 봉은사로 68길 41   
..      ...          ...                        ...   
541      93       컨트리사이드  경기도 고양시 덕양구 행당로 1 4층 401호   
543      48   식물학 커피로스터스        서울시 강남구 강남대로154길 33   
544      49      싸인포스트커피           서울시 동작구 솔밭로 97-6   
553      61         오드이븐    부산 중구 중구로29번길 7 1층 오드이븐   
558      36  보사노바 커피로스터스          강원 강릉시 창해로14번길 28   

                                     제조원 및 소재지  
0                 (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  
3     뉴웨이브커피로스터스, 서울시 양천구 오목로 56길 2, 영원빌딩 103호  
4          베르크로스터스, 부산광역시 사상구 학감대로222번길 83, 2층  
6    라메르판지 비앤알 / 울산광역시 북구 판지1길 30 (지상2~3층 구유동)  
7                      서울시 강남구 봉은사로 68길 41, 1층  
..                                         ...  
541                   경기도 고양시 덕양구 행당로1 4층 401호  
543         식물학커피로스터스, 서울특별시 용산구 서빙고로67길 21 2층  
544                  싸인포스트커피, 서울시 동작구 솔밭로 97-6  
553                   부산광역시 중구 중구로 29번길 7 오드이븐  
558            보사노바로스팅팩토리, 강원도 강릉시 경강로2660번길 7  

[318 rows x 4 columns]

<로스터리 주소와 제조원이 항상 일치하지는 않는다.> <br>
로스터리 주소 : 커피네츄럴스페셜티, 헤베커피 <br>
제조원은 그냥 0으로 둬야할 것 같다.

In [69]:
df.loc[df['로스터리'] == '커피네츄럴스페셜티', '로스터리 주소'] = df['제조원 및 소재지'][df['로스터리'] == '커피네츄럴스페셜티'].values[0]
df.loc[df['로스터리'] == '헤베커피', '로스터리 주소'] = df['제조원 및 소재지'][df['로스터리'] == '헤베커피'].values[0]

In [70]:
df['제조원 및 소재지'][df['로스터리'] == '커피네츄럴스페셜티'].values[0]

'경기도 고양시 덕양구 행신동 741-17 1층'

In [71]:
df['로스터리 주소'][df['로스터리']=='커피네츄럴스페셜티']

17     경기도 고양시 덕양구 행신동 741-17 1층
59     경기도 고양시 덕양구 행신동 741-17 1층
248    경기도 고양시 덕양구 행신동 741-17 1층
322    경기도 고양시 덕양구 행신동 741-17 1층
527    경기도 고양시 덕양구 행신동 741-17 1층
Name: 로스터리 주소, dtype: object

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 559 non-null    int64 
 1   이름                 559 non-null    object
 2   로스터리               559 non-null    object
 3   타입                 559 non-null    object
 4   로스팅 포인트            559 non-null    object
 5   지속가능성              559 non-null    object
 6   컵 노트               559 non-null    object
 7   바디감                559 non-null    int64 
 8   신맛                 559 non-null    int64 
 9   단맛                 559 non-null    int64 
 10  쓴맛                 559 non-null    int64 
 11  커피 소개              559 non-null    object
 12  나라                 559 non-null    object
 13  로스터리 소개            559 non-null    object
 14  로스터리 주소            559 non-null    object
 15  이름.1               559 non-null    object
 16  식품의 유형             554 non-null    object
 1

In [73]:
# 업데이트 wondoofin.csv
df.to_csv('wondoofin.csv', index=False)

***
나라 전처리
***

In [3]:
df = pd.read_csv('wondoofin.csv')
df.head(2)

id            이름   로스터리    타입 로스팅 포인트 지속가능성                 컵 노트  바디감  \
0    43         데일리스윗  언더프레셔   블렌드   미디엄다크     0  ['토피', '맥아', '흑설탕']    4   
1  3896  브라질 세라도 디카페인   운조커피  디카페인   미디엄다크     0             ['null']    4   

   신맛  단맛  ...                     제조원 및 소재지         유통기한        제조일자  \
0   2   4  ...  (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  제조일로부터 12개월      상단 표기일   
1   2   4  ...        경기도 화성시 동탄지성로 306번길 68    제조일로부터 1년  주문 확인 후 제조   

          내용량                                         보관 방법  \
0  140g / 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.   
1    500g 1kg       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관   

                          원재료 및 함량   제품문의 관련 주소 및 전화번호 카페인 로스터리ID  price  
0  커피원두 100% (브라질산 60%, 과테말라산 40%)  핸디엄 고객센터 1599-2681   1     57  11800  
1                        커피원두 100%  운조커피 010-4535-8819   0     67  13000  

[2 rows x 27 columns]

In [144]:
df['나라'] = df['나라'].apply(literal_eval)
# for i in range(len(df)):
#     for k in range(len(df['나라'][i])):
#         df['나라'][i][k] = df['나라'][i][k].lstrip() # 띄어쓰기가 단어 첫번째에 있으면 지운다!
df['나라'] = [[country.lstrip() for country in sublist] if sublist != 0 else sublist for sublist in df['나라']]

df['나라'][0]

['브라질', '과테말라']

In [145]:
df['id'][df['나라']==0]

11     3460
15     3907
30     4717
67       61
72     4702
167    3232
179    1707
192    3177
218    3196
242    4750
301    3438
381    4817
416    4707
474    1038
548    4643
554    3231
Name: id, dtype: int64

In [146]:
# Flatten the 'country' column and remove 0 values
flat_countries = [subitem for sublist in df['나라'] if sublist != 0 for subitem in sublist]

# Get unique country names
unique_countries = list(set(flat_countries))

In [147]:
len(unique_countries)

130

In [148]:
unique_countries[41:50]

['산 로렌조', '탄자이나', '칼다스', '가차타', '산타아나', '안티오키아 볼리바르', '카우카', '괴템말라', '라 단타']

***
가격 정보
***

In [149]:
price_df = pd.read_csv('price.csv')
price_df.head()

Name    Price      Roastery
0            진양조  10,000원         스탠딩커피
1      클래식 에스프레소  13,000원  리플렉트 커피 로스터스
2  디카페인 올 데이 블렌드  10,600원         디카커피랩
3         우리집블렌드   9,400원          챁챁커피
4        오닉스 블렌드  12,000원        필리그리커피

In [150]:
price_df.shape

(345, 3)

In [151]:
len(price_df['Name'].unique())

335

In [152]:
price_df[price_df['Name'].duplicated()]

Name    Price      Roastery
70   디카페인 콜롬비아  16,000원       커피가게 동경
97    디카페인 블렌드  12,000원     하우트 커피컴퍼니
98   콜롬비아 디카페인  12,800원         언더프레셔
101  콜롬비아 디카페인  15,000원     피어 커피로스터스
142     피크 블렌드  10,000원          오드이븐
159  콜롬비아 디카페인  15,000원         스탠딩커피
205  콜롬비아 디카페인  11,000원         하늘에커피
218  콜롬비아 디카페인  15,000원    502 커피로스터스
271   브라질 디카페인   9,100원  리플로우 커피 로스터스
284    클래식 블렌드  12,000원    502 커피로스터스

In [153]:
price_df[price_df['Name'] == '디카페인 블렌드']

Name    Price   Roastery
26  디카페인 블렌드  15,000원    웨이브온 커피
97  디카페인 블렌드  12,000원  하우트 커피컴퍼니

In [154]:
df[(df['이름'] == '디카페인 블렌드') & (df['로스터리']=='웨이브온 커피')]

id        이름     로스터리    타입 로스팅 포인트 지속가능성  \
532  226  디카페인 블렌드  웨이브온 커피  디카페인      다크  직접무역   

                                    컵 노트  바디감  신맛  단맛  ...  식품의 유형  \
532  ['견과류', '카라멜', '아몬드', '카카오', '황설탕']    4   1   3  ...      커피   

                          제조원 및 소재지       유통기한   제조일자         내용량  \
532  웨이브온커피, 부산광역시 기장군 장안읍 해맞이로 286  제조일로부터 1년  별도 표시  200g / 1kg   

                    보관 방법  원재료 및 함량      제품문의 관련 주소 및 전화번호 카페인 로스터리ID  
532  밀봉 후 직사광선을 피하여 실온 보관  커피원두100%  코케 고객센터 070-4647-1868   0     69  

[1 rows x 26 columns]

In [155]:
df[(df['이름'] == price_df['Name'][26]) & (df['로스터리'] == price_df['Roastery'][26])]

id        이름     로스터리    타입 로스팅 포인트 지속가능성  \
532  226  디카페인 블렌드  웨이브온 커피  디카페인      다크  직접무역   

                                    컵 노트  바디감  신맛  단맛  ...  식품의 유형  \
532  ['견과류', '카라멜', '아몬드', '카카오', '황설탕']    4   1   3  ...      커피   

                          제조원 및 소재지       유통기한   제조일자         내용량  \
532  웨이브온커피, 부산광역시 기장군 장안읍 해맞이로 286  제조일로부터 1년  별도 표시  200g / 1kg   

                    보관 방법  원재료 및 함량      제품문의 관련 주소 및 전화번호 카페인 로스터리ID  
532  밀봉 후 직사광선을 피하여 실온 보관  커피원두100%  코케 고객센터 070-4647-1868   0     69  

[1 rows x 26 columns]

In [156]:
price_df['Price'][26]

'15,000원'

In [177]:
df['price'] = 0
for i in range(len(price_df)):
    # df['price'][(df['이름'] == price_df['Name'][i]) & (df['로스터리'] == price_df['Roastery'][i])] = price_df['Price'][i]
    mask = (df['이름'] == price_df['Name'][i]) & (df['로스터리'] == price_df['Roastery'][i])
    price_value = price_df['Price'][i]

    # Convert 'price_value' to an integer (assuming it's a numeric value)
    price_value = int(price_value.replace(',', '').replace('원', ''))

    df.loc[mask, 'price'] = price_value

In [178]:
df['price'][(df['이름'] == price_df['Name'][26]) & (df['로스터리'] == price_df['Roastery'][26])]

532    15000
Name: price, dtype: int64

In [191]:
df.head(2)

id            이름   로스터리    타입 로스팅 포인트 지속가능성                 컵 노트  바디감  \
0    43         데일리스윗  언더프레셔   블렌드   미디엄다크     0  ['토피', '맥아', '흑설탕']    4   
1  3896  브라질 세라도 디카페인   운조커피  디카페인   미디엄다크     0             ['null']    4   

   신맛  단맛  ...                     제조원 및 소재지         유통기한        제조일자  \
0   2   4  ...  (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  제조일로부터 12개월      상단 표기일   
1   2   4  ...        경기도 화성시 동탄지성로 306번길 68    제조일로부터 1년  주문 확인 후 제조   

          내용량                                         보관 방법  \
0  140g / 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.   
1    500g 1kg       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관   

                          원재료 및 함량   제품문의 관련 주소 및 전화번호 카페인 로스터리ID  price  
0  커피원두 100% (브라질산 60%, 과테말라산 40%)  핸디엄 고객센터 1599-2681   1     57  11800  
1                        커피원두 100%  운조커피 010-4535-8819   0     67      0  

[2 rows x 27 columns]

In [192]:
df['price'].describe()

count      559.000000
mean      8015.921288
std       8202.946439
min          0.000000
25%          0.000000
50%       9700.000000
75%      14000.000000
max      46000.000000
Name: price, dtype: float64

In [193]:
# Calculate the second most frequent value of the 'price' column. most frequent value is 0
mode_value = df['price'].mode().iloc[0]
df_without_mode = df[df['price'] != mode_value]

# Calculate the mode again, which is the second most frequent value
second_mode_value = df_without_mode['price'].mode().iloc[0]
second_mode_value

# Replace 0 values with NaN so that we can calculate the mean without them
df['price'].replace(0, pd.NA, inplace=True)

# Fill NaN values (including the former 0 values) with the mean value
df['price'].fillna(second_mode_value, inplace=True)

In [194]:
df['price'].describe()

count      559.000000
mean     13806.618962
std       4013.491028
min       5500.000000
25%      13000.000000
50%      13000.000000
75%      14000.000000
max      46000.000000
Name: price, dtype: float64

In [195]:
# 업데이트 wondoofin.csv
df.to_csv('wondoofin.csv', index=False)

In [196]:
df.head(2)

id            이름   로스터리    타입 로스팅 포인트 지속가능성                 컵 노트  바디감  \
0    43         데일리스윗  언더프레셔   블렌드   미디엄다크     0  ['토피', '맥아', '흑설탕']    4   
1  3896  브라질 세라도 디카페인   운조커피  디카페인   미디엄다크     0             ['null']    4   

   신맛  단맛  ...                     제조원 및 소재지         유통기한        제조일자  \
0   2   4  ...  (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  제조일로부터 12개월      상단 표기일   
1   2   4  ...        경기도 화성시 동탄지성로 306번길 68    제조일로부터 1년  주문 확인 후 제조   

          내용량                                         보관 방법  \
0  140g / 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.   
1    500g 1kg       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관   

                          원재료 및 함량   제품문의 관련 주소 및 전화번호 카페인 로스터리ID  price  
0  커피원두 100% (브라질산 60%, 과테말라산 40%)  핸디엄 고객센터 1599-2681   1     57  11800  
1                        커피원두 100%  운조커피 010-4535-8819   0     67  13000  

[2 rows x 27 columns]

In [198]:
# value = 10000  # Example integer
# formatted_value = "{:,}".format(value)
# formatted_value

'10,000'

***
가격 정보 coffeeDB에 넣기
***

In [132]:
pricedf = pd.read_csv('wondoofin.csv')
coffeeDB = pd.read_csv('mysite/data/coffeeDB.csv', index_col=0)
pricedf = pricedf[['id', 'price']]

# coffeeDB = pd.merge(coffeeDB, pricedf, on="id")
coffeeDB.head(2)

이름  로스터리ID                                              커피 소개  \
id                                                                              
43           데일리스윗      57  중남미 블렌딩으로 깊이 있는 단맛과 초콜렛티함이 조화롭게 어우러진 하우스 블렌드 입...   
3896  브라질 세라도 디카페인      67  땅콩의 고소함과 코코아 파우더의 쌉싸름함 진득한 단맛과 클래식한 커피향  중강배전으...   

        타입 로스팅 포인트 지속가능성                   컵 노트  바디감  신맛  단맛  ...         나라  \
id                                                            ...              
43     블렌드   미디엄다크   NaN  ['토피', ' 맥아', ' 흑설탕']    4   2   4  ...  브라질, 과테말라   
3896  디카페인   미디엄다크   NaN                    NaN    4   2   4  ...        세라도   

        식품의 유형                     제조원 및 소재지         유통기한        제조일자  \
id                                                                      
43    커피(볶은커피)  (주)핸디엄, 경기도 여주시 세종대왕면 상동길 68  제조일로부터 12개월      상단 표기일   
3896        커피        경기도 화성시 동탄지성로 306번길 68    제조일로부터 1년  주문 확인 후 제조   

             내용량                                         보관 방법  \
id                                                               
43    140g / 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.   
3896    500g 1kg       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관   

                             원재료 및 함량   제품문의 관련 주소 및 전화번호  price  
id                                                                
43    커피원두 100% (브라질산 60%, 과테말라산 40%)  핸디엄 고객센터 1599-2681  11800  
3896                        커피원두 100%  운조커피 010-4535-8819  13000  

[2 rows x 22 columns]

In [131]:
coffeeDB.to_csv('mysite/data/coffeeDB.csv', index=False)

***
로스터리 전화번호 
***

In [8]:
import pandas as pd
df = pd.read_csv('wondoofin.csv')
roastery = df[['로스터리ID', '로스터리', '로스터리 주소', '제품문의 관련 주소 및 전화번호']]
roastery

로스터리ID         로스터리                                         로스터리 주소  \
0        57        언더프레셔                           서울 강남구 영동대로106길 37 1층   
1        67         운조커피                          경기도 화성시 동탄지성로 306번길 68   
2        77    제이엠디커피팩토리  경기 고양시 일산동구 장대길 75-53 나동 1층 JMD Coffee Factory   
3        11  뉴웨이브 커피로스터스                           서울특별시 양천구 오목로56길 2,1층   
4        34     베르크 로스터스                          부산 부산진구 전포동 356-6 지하1층   
..      ...          ...                                             ...   
554       4   UFO 커피로스터스                                  서울시 관악구 은천로 39   
555     109          핸디엄                           서울특별시 강남구 영동대로106길 37   
556      30   링크 커피 로스터스                    전라남도 장성군 진원면 나노산단2로 52, 205호   
557      82      카페소소한사치                             경기도 부천시 소사구 호현로 489   
558      36  보사노바 커피로스터스                               강원 강릉시 창해로14번길 28   

            제품문의 관련 주소 및 전화번호  
0          핸디엄 고객센터 1599-2681  
1          운조커피 010-4535-8819  
2     제이엠디커피팩토리 010-9903-8794  
3       코케 고객센터 070-4647-1868  
4       코케 고객센터 070-4647-1868  
..                        ...  
554  UFO 커피로스터스 010-4240-8455  
555         핸디엄 010-4194-8939  
556  링크 커피 로스터스 010-5035-1425  
557     카페소소한사치 010-9340-4364  
558     코케 고객센터 070-4647-1868  

[559 rows x 4 columns]

In [9]:
len(roastery['로스터리ID'].unique())

114

In [10]:
roastery['제품문의 관련 주소 및 전화번호'][roastery['로스터리ID'] == 0].unique().tolist()

['코케 고객센터 070-4647-1868', '180커피로스터스 010-4510-8597']

In [11]:
koke = roastery[roastery['제품문의 관련 주소 및 전화번호'] == '코케 고객센터 070-4647-1868']
len(koke['로스터리ID'].unique())

54

In [12]:
rdict = {}
id_list = []
for i in range(len(df)):
    rdict[roastery['로스터리ID'][i]] = []
    
    id_list.append(roastery['로스터리ID'][i])
    id_list = list(set(id_list))

len(id_list)

114

In [13]:
for id in id_list:
    ad = roastery['제품문의 관련 주소 및 전화번호'][roastery['로스터리ID'] == id].unique().tolist()
    rdict[id] = ad
rdict

{57: ['핸디엄 고객센터 1599-2681', '언더프레셔 1599-2681'],
 67: ['운조커피 010-4535-8819'],
 77: ['제이엠디커피팩토리 010-9903-8794'],
 11: ['코케 고객센터 070-4647-1868'],
 34: ['코케 고객센터 070-4647-1868'],
 78: ['조바티커피 070-4203-6900', '조바티커피'],
 18: ['라메르판지비앤알 0522928787', '라메르판지 비앤알 0522928787'],
 98: ['코케 고객센터 070-4647-1868'],
 29: ['리플로우 커피 로스터스 010-4932-9198', '코케 고객센터 070-4647-1868'],
 85: ['커스프 커피 010-3267-0430', '커스프 커피 070-8098-8025'],
 8: ['코케 고객센터 070-4647-1868'],
 75: ['전광수 커피 010-2936-0501'],
 63: ['오프사이트 0507-1399-2781'],
 43: ['수수커피 010-5068-4771', '코케 고객센터 070-4647-1868'],
 6: ['그린마일커피'],
 89: ['커피네츄럴스페셜티 01090945303', '커피네츄럴스페셜티'],
 101: ['[폰트오피스] 031-946-8549', '폰트커피 010-9483-1853'],
 102: ['코케 고객센터 070-4647-1868'],
 0: ['코케 고객센터 070-4647-1868', '180커피로스터스 010-4510-8597'],
 104: ['피어 커피로스터스 010-4501-1464',
  '피어 커피로스터스 010-9388-7053',
  '코케 고객센터 070-4647-1868'],
 113: ['코케 고객센터 070-4647-1868'],
 46: ['코케 고객센터 070-4647-1868', '스티머스 커피팩토리 010-6620-5557'],
 13: ['디그리 01031660563'],
 1: ['코케 고객센터 070-46

In [61]:
# Create a new dictionary to store the filtered lists
# 전화 번호 정보가 2개 이상일때 그 중 하나가 코케 정보라면 코케 정보 삭제, 전화번호 정보가 한개인데 코케라면 남겨두기
filtered_dict = {}

for key, value in rdict.items():
    if len(value) > 2:
        value = [item for item in value if item != '코케 고객센터 070-4647-1868']
    filtered_dict[key] = value

print(filtered_dict)

{57: ['핸디엄 고객센터 1599-2681', '언더프레셔 1599-2681'], 67: ['운조커피 010-4535-8819'], 77: ['제이엠디커피팩토리 010-9903-8794'], 11: ['코케 고객센터 070-4647-1868'], 34: ['코케 고객센터 070-4647-1868'], 78: ['조바티커피 070-4203-6900', '조바티커피'], 18: ['라메르판지비앤알 0522928787', '라메르판지 비앤알 0522928787'], 98: ['코케 고객센터 070-4647-1868'], 29: ['리플로우 커피 로스터스 010-4932-9198', '코케 고객센터 070-4647-1868'], 85: ['커스프 커피 010-3267-0430', '커스프 커피 070-8098-8025'], 8: ['코케 고객센터 070-4647-1868'], 75: ['전광수 커피 010-2936-0501'], 63: ['오프사이트 0507-1399-2781'], 43: ['수수커피 010-5068-4771', '코케 고객센터 070-4647-1868'], 6: ['그린마일커피'], 89: ['커피네츄럴스페셜티 01090945303', '커피네츄럴스페셜티'], 101: ['[폰트오피스] 031-946-8549', '폰트커피 010-9483-1853'], 102: ['코케 고객센터 070-4647-1868'], 0: ['코케 고객센터 070-4647-1868', '180커피로스터스 010-4510-8597'], 104: ['피어 커피로스터스 010-4501-1464', '피어 커피로스터스 010-9388-7053'], 113: ['코케 고객센터 070-4647-1868'], 46: ['코케 고객센터 070-4647-1868', '스티머스 커피팩토리 010-6620-5557'], 13: ['디그리 01031660563'], 1: ['코케 고객센터 070-4647-1868'], 83: ['카페플롬 010-4155-8282'], 56: ['010-513

In [28]:
filtered_dict[18]

['라메르판지비앤알 0522928787', '라메르판지 비앤알 0522928787']

In [31]:
# 전화번호 정보만 남기기
phone_numbers_data = {}

for key, value in filtered_dict.items():
    phone_numbers = []
    for item in value:
        # Use regular expressions to extract phone numbers
        numbers_and_hyphens = re.findall(r'[0-9-]+', item)
        # Join the extracted parts into a single string
        phone_number = ''.join(numbers_and_hyphens)
        if len(phone_number) > 0:
            phone_numbers.append(phone_number)
        phone_numbers = list(set(phone_numbers))
    phone_numbers_data[key] = phone_numbers

print(phone_numbers_data)

{57: ['1599-2681'], 67: ['010-4535-8819'], 77: ['010-9903-8794'], 11: ['070-4647-1868'], 34: ['070-4647-1868'], 78: ['070-4203-6900'], 18: ['0522928787'], 98: ['070-4647-1868'], 29: ['010-4932-9198', '070-4647-1868'], 85: ['070-8098-8025', '010-3267-0430'], 8: ['070-4647-1868'], 75: ['010-2936-0501'], 63: ['0507-1399-2781'], 43: ['070-4647-1868', '010-5068-4771'], 6: [], 89: ['01090945303'], 101: ['031-946-8549', '010-9483-1853'], 102: ['070-4647-1868'], 0: ['180010-4510-8597', '070-4647-1868'], 104: ['010-4501-1464', '010-9388-7053'], 113: ['070-4647-1868'], 46: ['010-6620-5557', '070-4647-1868'], 13: ['01031660563'], 1: ['070-4647-1868'], 83: ['010-4155-8282'], 56: ['010-5134-1306'], 24: ['010-2942-3069'], 47: ['070-4647-1868', '010-2031-5959'], 20: ['01089630224'], 15: ['070-4647-1868', '010-3102-1655'], 62: ['929-3106', '07080283535', '010-8720-3414'], 100: ['070-4647-1868'], 7: ['01023160961'], 69: ['051-727-16602', '070-4647-1868'], 49: ['070-4647-1868', '010-4879-1791'], 70: ['0

In [65]:
# Filter rows where 'phone_numbers' contain a number in "0000-0000" format
pattern = r'\d{4}-\d{4}'
filtered_df = phone_df[phone_df['phone_numbers'].apply(lambda x: any(re.match(pattern, number) for number in x))]

# Print the 'id' values for the filtered rows
id_values = filtered_df['id'].tolist()
print(id_values)

[57, 63, 109]


위의 코드로 전화번호를 추출한 결과 주소내의 숫자와 -로 전화번호를 만들어버렸다.
- 다시ㅠㅠㅠ

In [79]:
def filter_phone_numbers(phone_numbers):
    filtered_numbers = []
    for number in phone_numbers:
        # If there is no hyphen, it must be longer than 7
        if '-' not in number and len(number) > 7:
            filtered_numbers.append(number)
        # If there is a hyphen, check the length of parts before and after the hyphen
        elif '-' in number:
            parts = number.split('-')
            if len(parts) >= 2 and len(parts[0]) >= 2 and len(parts[1]) >= 2:
                filtered_numbers.append(number)
    return filtered_numbers

phone_numbers_data = {}

for key, value in filtered_dict.items():
    phone_numbers = []
    for item in value:
        # Use a regular expression to extract numbers and hyphens without alphabetic characters in between
        phone_numbers_found = re.findall(r'[\d-]+', item)
        phone_numbers.extend(phone_numbers_found)
    
    # Filter the extracted phone numbers
    filtered_numbers = filter_phone_numbers(phone_numbers)
    
    if filtered_numbers:
        phone_numbers_data[key] = list(set(filtered_numbers))
    else:
        phone_numbers_data[key] = []

print(phone_numbers_data)

{57: ['1599-2681'], 67: ['010-4535-8819'], 77: ['010-9903-8794'], 11: ['070-4647-1868'], 34: ['070-4647-1868'], 78: ['070-4203-6900'], 18: ['0522928787'], 98: ['070-4647-1868'], 29: ['010-4932-9198', '070-4647-1868'], 85: ['070-8098-8025', '010-3267-0430'], 8: ['070-4647-1868'], 75: ['010-2936-0501'], 63: ['0507-1399-2781'], 43: ['070-4647-1868', '010-5068-4771'], 6: [], 89: ['01090945303'], 101: ['031-946-8549', '010-9483-1853'], 102: ['070-4647-1868'], 0: ['010-4510-8597', '070-4647-1868'], 104: ['010-4501-1464', '010-9388-7053'], 113: ['070-4647-1868'], 46: ['010-6620-5557', '070-4647-1868'], 13: ['01031660563'], 1: ['070-4647-1868'], 83: ['010-4155-8282'], 56: ['010-5134-1306'], 24: ['010-2942-3069'], 47: ['070-4647-1868', '010-2031-5959'], 20: ['01089630224'], 15: ['070-4647-1868', '010-3102-1655'], 62: ['010-8720-3414', '07080283535'], 100: ['070-4647-1868'], 7: ['01023160961'], 69: ['051-727-1660', '070-4647-1868'], 49: ['070-4647-1868', '010-4879-1791'], 70: ['010-6434-8150'], 

In [80]:
# Create an empty list to hold data for the DataFrame
data = []

for key, values in phone_numbers_data.items():
    data.append([key, values])

# Create a DataFrame from the list of data
phone_df = pd.DataFrame(data, columns=['id', 'phone_numbers'])

# Display the DataFrame
print(phone_df)

      id               phone_numbers
0     57                 [1599-2681]
1     67             [010-4535-8819]
2     77             [010-9903-8794]
3     11             [070-4647-1868]
4     34             [070-4647-1868]
..   ...                         ...
109   39              [02-2643-2988]
110   65             [010-3309-3120]
111  109  [1599-2681, 010-4194-8939]
112  103             [070-4647-1868]
113   14             [070-4647-1868]

[114 rows x 2 columns]


In [81]:
# Filter rows where 'phone_numbers' contain a number in "0000-0000" format
pattern = r'\d{4}-\d{4}'
filtered_df = phone_df[phone_df['phone_numbers'].apply(lambda x: any(re.match(pattern, number) for number in x))]

# Print the 'id' values for the filtered rows
id_values = filtered_df['id'].tolist()
print(id_values)

[57, 63, 109]


In [82]:
print(rdict[57])
print(rdict[63])
print(rdict[109])

['핸디엄 고객센터 1599-2681', '언더프레셔 1599-2681']
['오프사이트 0507-1399-2781']
['핸디엄 010-4194-8939', '핸디엄 1599-2681']


In [103]:
# Flatten the list of lists
flattened_v = [item for sublist in v for item in sublist]

# Check for duplicates using a set
duplicates = [item for item in set(flattened_v) if flattened_v.count(item) > 1]

if duplicates:
    print("Duplicates found:")
    for duplicate in duplicates:
        print(duplicate)
else:
    print("No duplicates found.")

Duplicates found:
1599-2681
070-4647-1868


In [111]:
def find_keys_by_value(input_dict, target_value):
    keys = []
    for key, value in input_dict.items():
        if value == target_value:
            keys.append(key)
    return keys

# Example usage:
target_value = ['1599-2681']
keys_with_target_value = find_keys_by_value(phone_numbers_data, target_value)
print(keys_with_target_value)

[57]


In [108]:
# 57, 109 : 핸디엄이랑 언더프레셔는 같은 회사!!
roastery[roastery['로스터리ID']==57]

로스터리ID   로스터리                로스터리 주소   제품문의 관련 주소 및 전화번호
0        57  언더프레셔  서울 강남구 영동대로106길 37 1층  핸디엄 고객센터 1599-2681
65       57  언더프레셔  서울 강남구 영동대로106길 37 1층  핸디엄 고객센터 1599-2681
286      57  언더프레셔  서울 강남구 영동대로106길 37 1층  핸디엄 고객센터 1599-2681
291      57  언더프레셔  서울 강남구 영동대로106길 37 1층     언더프레셔 1599-2681
415      57  언더프레셔  서울 강남구 영동대로106길 37 1층  핸디엄 고객센터 1599-2681

In [114]:
phone_df.shape

(114, 2)

In [120]:
phone_df = phone_df.rename(columns={'id':'로스터리ID'})
phone_df.head()

로스터리ID    phone_numbers
0      57      [1599-2681]
1      67  [010-4535-8819]
2      77  [010-9903-8794]
3      11  [070-4647-1868]
4      34  [070-4647-1868]

In [116]:
# phone_df랑 roasteryDB.csv랑 id로 합치기
DB = pd.read_csv('mysite/data/roasteryDB.csv', index_col=0)
DB.head()

로스터리ID         로스터리                                         로스터리 주소  \
0      57        언더프레셔                           서울 강남구 영동대로106길 37 1층   
1      67         운조커피                          경기도 화성시 동탄지성로 306번길 68   
2      77    제이엠디커피팩토리  경기 고양시 일산동구 장대길 75-53 나동 1층 JMD Coffee Factory   
3      11  뉴웨이브 커피로스터스                           서울특별시 양천구 오목로56길 2,1층   
4      34     베르크 로스터스                          부산 부산진구 전포동 356-6 지하1층   

                                             로스터리 소개  
0  “일상 속 특별한 휴식” 커피를 추출하는 기본 원리인 물줄기의 압력(Pressure...  
1  커피가 나에게 준 커다란 변화를 모두와 나누고 싶습니다. 나의 이름이 붙은 커피를 ...  
2  “JMD만의 명품 원두로 그 이상의 무한함을 창조하고자 합니다.” 커피는 이제 단지...  
3  뉴웨이브 커피로스터스는 확고한 커피의 철학으로 우리만의 스타일로 많은 커피를 전달하...  
4  부산에 위치한 스몰배치 스페셜티 커피 로스터리이자, 서로 다른 경험과 개성을 가진 ...

In [117]:
DB.shape

(114, 4)

In [121]:
roasteryDB = pd.merge(DB, phone_df, on="로스터리ID")
roasteryDB.head()

로스터리ID         로스터리                                         로스터리 주소  \
0      57        언더프레셔                           서울 강남구 영동대로106길 37 1층   
1      67         운조커피                          경기도 화성시 동탄지성로 306번길 68   
2      77    제이엠디커피팩토리  경기 고양시 일산동구 장대길 75-53 나동 1층 JMD Coffee Factory   
3      11  뉴웨이브 커피로스터스                           서울특별시 양천구 오목로56길 2,1층   
4      34     베르크 로스터스                          부산 부산진구 전포동 356-6 지하1층   

                                             로스터리 소개    phone_numbers  
0  “일상 속 특별한 휴식” 커피를 추출하는 기본 원리인 물줄기의 압력(Pressure...      [1599-2681]  
1  커피가 나에게 준 커다란 변화를 모두와 나누고 싶습니다. 나의 이름이 붙은 커피를 ...  [010-4535-8819]  
2  “JMD만의 명품 원두로 그 이상의 무한함을 창조하고자 합니다.” 커피는 이제 단지...  [010-9903-8794]  
3  뉴웨이브 커피로스터스는 확고한 커피의 철학으로 우리만의 스타일로 많은 커피를 전달하...  [070-4647-1868]  
4  부산에 위치한 스몰배치 스페셜티 커피 로스터리이자, 서로 다른 경험과 개성을 가진 ...  [070-4647-1868]

In [122]:
# 업데이트 mysite/data/roasteryDB.csv
roasteryDB.to_csv('mysite/data/roasteryDB.csv', index=False)

In [125]:
roasteryDB[roasteryDB['로스터리ID']==104]

로스터리ID       로스터리                  로스터리 주소  \
19     104  피어 커피로스터스  서울특별시 성동구 광나루로4가길 28 1층   

                                              로스터리 소개  \
19  PEER : 당신의 삶속에 더욱 더 가까이 동료,친구 라는 의미를 가지고 있는 PE...   

                     phone_numbers  
19  [010-4501-1464, 010-9388-7053]

***
컵 노트 카테고리 칼럼
***

In [7]:
category = pd.read_csv('category.csv')
category = category.rename(columns={'향료_풍미':'향료'})

# update category.csv
category.to_csv('category.csv', index=False)
test = pd.read_csv('category.csv')
test.head(2)

id                        컵 노트  꽃  과일  허브  달콤함  고소함  향료_풍미  초콜릿
0    43         ['토피', '맥아', '흑설탕']  0   0   0    1    0      1    0
1  3896                    ['null']  0   0   0    0    0      0    0
2  3458                    ['null']  0   0   0    0    0      0    0
3    22  ['다크 초콜릿', '마카다미아', '카라멜']  0   0   0    1    1      0    1
4   180               ['초콜릿', '과일']  0   1   0    0    0      0    1

In [14]:
features = pd.read_csv('mysite/data/features.csv')
features.head()

id  바디감  신맛  단맛  쓴맛  타입_디카페인  타입_블렌드  타입_싱글오리진  지속가능성_0  지속가능성_공정무역  ...  \
0    43    4   2   4   3        0       1         0        1           0  ...   
1  3896    4   2   4   3        1       0         0        1           0  ...   
2  3458    3   4   5   1        0       0         1        0           0  ...   
3    22    4   1   3   4        0       1         0        0           1  ...   
4   180    5   3   5   1        0       0         1        1           0  ...   

   로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  달콤함  고소함  향료_풍미  초콜릿  
0               0            0              1  0   0   0    1    0      1    0  
1               0            0              1  0   0   0    0    0      0    0  
2               1            0              0  0   0   0    0    0      0    0  
3               0            0              0  0   0   0    1    1      0    1  
4               0            0              1  0   1   0    0    0      0    1  

[5 rows x 24 columns]

In [15]:
features.columns

Index(['id', '바디감', '신맛', '단맛', '쓴맛', '타입_디카페인', '타입_블렌드', '타입_싱글오리진',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료_풍미', '초콜릿'],
      dtype='object')

In [17]:
column_mapping = {'바디감' : 'body', '신맛' : 'sour', '단맛' : 'sweet', '쓴맛' : 'bitter', '타입_디카페인' : 'decaf', '타입_블렌드' : 'blend', '향료_풍미':'향료'}
features.rename(columns=column_mapping, inplace=True)
features.columns

Index(['id', 'body', 'sour', 'sweet', 'bitter', 'decaf', 'blend', '타입_싱글오리진',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [18]:
# update features.csv
features.to_csv('mysite/data/features.csv', index=False)
test = pd.read_csv('mysite/data/features.csv')
test.head(2)

id  body  sour  sweet  bitter  decaf  blend  타입_싱글오리진  지속가능성_0  \
0    43     4     2      4       3      0      1         0        1   
1  3896     4     2      4       3      1      0         0        1   

   지속가능성_공정무역  ...  로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  \
0           0  ...               0            0              1  0   0   0   
1           0  ...               0            0              1  0   0   0   

   달콤함  고소함  향료  초콜릿  
0    1    0   1    0  
1    0    0   0    0  

[2 rows x 24 columns]

***
디카페인
****

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('wondoofin.csv')
df.columns

Index(['id', '이름', '로스터리', '타입', '로스팅 포인트', '지속가능성', '컵 노트', '바디감', '신맛', '단맛',
       '쓴맛', '커피 소개', '나라', '로스터리 소개', '로스터리 주소', '이름.1', '식품의 유형',
       '제조원 및 소재지', '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량',
       '제품문의 관련 주소 및 전화번호', '카페인', '로스터리ID', 'price'],
      dtype='object')

In [4]:
features = pd.read_csv('features.csv')
features.columns

Index(['id', '바디감', '신맛', '단맛', '쓴맛', '타입_디카페인', '타입_블렌드', '타입_싱글오리진',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [5]:
fe = pd.read_csv('mysite/data/features.csv')
fe.columns

Index(['id', 'body', 'sour', 'sweet', 'bitter', 'decaf', 'blend', '타입_싱글오리진',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [7]:
type_c = features[['id', '타입_디카페인', '타입_블렌드', '타입_싱글오리진']]
type_c.head()

id  타입_디카페인  타입_블렌드  타입_싱글오리진
0    43        0       1         0
1  3896        1       0         0
2  3458        0       0         1
3    22        0       1         0
4   180        0       0         1

In [8]:
type_c[type_c['id']==4483]

id  타입_디카페인  타입_블렌드  타입_싱글오리진
186  4483        1       0         0

In [9]:
decaf = type_c[type_c['타입_디카페인'] == 1]

In [10]:
decaf.shape

(93, 4)

In [12]:
decaf = list(decaf['id'])
decaf

[3896,
 4603,
 3460,
 4737,
 4701,
 1055,
 127,
 1477,
 3180,
 4599,
 82,
 137,
 1084,
 61,
 194,
 3909,
 1505,
 37,
 4842,
 1495,
 1017,
 1083,
 3340,
 4878,
 4600,
 124,
 3914,
 1707,
 135,
 4483,
 3895,
 3069,
 3586,
 3333,
 4487,
 3148,
 1013,
 3196,
 3447,
 3614,
 4614,
 4408,
 1049,
 1296,
 3042,
 125,
 44,
 4686,
 4716,
 1045,
 3438,
 4017,
 189,
 1086,
 3075,
 3516,
 3921,
 116,
 4328,
 3963,
 3394,
 4172,
 136,
 2999,
 1714,
 282,
 3357,
 4707,
 3946,
 1573,
 1085,
 3508,
 1478,
 4601,
 1038,
 3058,
 3206,
 4407,
 3725,
 4602,
 4824,
 1088,
 3587,
 4449,
 1094,
 2995,
 3214,
 226,
 3897,
 4004,
 1105,
 4643,
 3231]

In [15]:
len(decaf)

93

In [13]:
import pickle
# Pickle dump the set of links to a file
with open("decaf.pkl", "wb") as file:
    pickle.dump(decaf, file)

In [18]:
# package import
import pandas as pd

from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.options import Options

import re
import time
# from tqdm import tqdm_notebook
from tqdm import tqdm

import pickle
import itertools

In [19]:
# Load the set of links from the file
with open("decaf.pkl", "rb") as file:
    loaded_decaf = pickle.load(file)

# Print the length loaded set of links. Intended length : 556
print("Intended length : 93 == ", len(loaded_decaf))

# change set to list
decaf_list = list(loaded_decaf)
print("list length : ", len(decaf_list))

Intended length : 93 ==  93
list length :  93


In [22]:
decaf_links = []
for item in decaf_list:
    link = 'https://biz.koke.kr/products/' + str(item)
    decaf_links.append(link)
    print(item, ' : ', link)

print(len(decaf_links) == len(decaf_list))

3896  :  https://biz.koke.kr/products/3896
4603  :  https://biz.koke.kr/products/4603
3460  :  https://biz.koke.kr/products/3460
4737  :  https://biz.koke.kr/products/4737
4701  :  https://biz.koke.kr/products/4701
1055  :  https://biz.koke.kr/products/1055
127  :  https://biz.koke.kr/products/127
1477  :  https://biz.koke.kr/products/1477
3180  :  https://biz.koke.kr/products/3180
4599  :  https://biz.koke.kr/products/4599
82  :  https://biz.koke.kr/products/82
137  :  https://biz.koke.kr/products/137
1084  :  https://biz.koke.kr/products/1084
61  :  https://biz.koke.kr/products/61
194  :  https://biz.koke.kr/products/194
3909  :  https://biz.koke.kr/products/3909
1505  :  https://biz.koke.kr/products/1505
37  :  https://biz.koke.kr/products/37
4842  :  https://biz.koke.kr/products/4842
1495  :  https://biz.koke.kr/products/1495
1017  :  https://biz.koke.kr/products/1017
1083  :  https://biz.koke.kr/products/1083
3340  :  https://biz.koke.kr/products/3340
4878  :  https://biz.koke.kr/

In [18]:
# return type, roasting point, and sustainability of the coffee
def other(driver):
    type = ['블렌드', '싱글오리진', '디카페인']
    roast = ['라이트', '라이트미디엄', '미디엄', '미디엄다크', '다크']
    # sustain = ['직접무역', '공정무역', '유기농']
    tmp = {'type' : [],
           'roast' : [],
           'sustain' : []}
    
    elements = driver.find_elements(By.CSS_SELECTOR, 'li.lg\:block')
    
    result = []
    for i in range(len(elements)):
        html = elements[i].get_attribute('innerHTML')
        # Parse the HTML content
        soup = BeautifulSoup(html, 'html.parser')
    
        if soup.find_all('svg', 'hidden'):
            result.append(elements[i].text)
    
    for i in result:
        if i in type:
            tmp['type'].append(i)
        elif i in roast:
            tmp['roast'].append(i)
        else: # i in sustain:
            tmp['sustain'].append(i)
    return tmp

In [23]:
service = Service(path= './driver/chromedriver.exe')
driver = webdriver.Chrome(service=service)
time.sleep(3)

data = {}
for link in tqdm(decaf_links):
    driver.get(link)
    time.sleep(3)
    
    # product id
    pattern = r"https://biz.koke.kr/products/(\d+)"  # Regular expression pattern to match the product number
    match = re.search(pattern, link)
    if match:
        product_number = match.group(1)
        data[product_number] = dict.fromkeys(['타입', '로스팅 포인트', '지속가능성'])  # add coffee names as keys and empty dictionary as values
        
    # others : '타입', '로스팅 포인트', '지속가능성'
    tmp = other(driver)
    for i in range(len(tmp)):
        keys = list(tmp.keys())
        if keys[i] == 'type':
            data[product_number]['타입'] = tmp['type']
        elif keys[i] == 'roast':
            data[product_number]['로스팅 포인트'] = tmp['roast']
        else:
            data[product_number]['지속가능성'] = tmp['sustain']
    time.sleep(3)
                
    
driver.quit()

time.sleep(3)

if len(data.keys()) == 93:
    print('{} out of 93'.format(len(data.keys())), 'Success!')
    df = pd.DataFrame.from_dict(data)
    newdf = df.T
    newdf.to_csv('decaf.csv')
else:
    print('{} out of 93'.format(len(data.keys())))
    # export current collected data
    comparedf = pd.DataFrame.from_dict(data)
    comparedf = comparedf.T
    comparedf.to_csv('compare.csv')

100%|██████████| 93/93 [10:30<00:00,  6.78s/it]


93 out of 93 Success!


In [38]:
decafdf = pd.read_csv('decaf.csv')
decafdf.head()

Unnamed: 0                 타입     로스팅 포인트     지속가능성
0        3896  ['싱글오리진', '디카페인']   ['미디엄다크']        []
1        4603  ['싱글오리진', '디카페인']   ['미디엄다크']  ['공정무역']
2        3460  ['싱글오리진', '디카페인']     ['미디엄']   ['유기농']
3        4737  ['싱글오리진', '디카페인']  ['라이트미디엄']        []
4        4701    ['블렌드', '디카페인']     ['미디엄']        []

In [39]:
from ast import literal_eval
# 리스트 처리
decafdf['타입'] = decafdf['타입'].apply(literal_eval)
decafdf['타입'][0]

['싱글오리진', '디카페인']

In [40]:
coffee_type = ['블렌드', '싱글오리진', '디카페인']
for i in coffee_type:
    decafdf[i] = 0
decafdf.head()

Unnamed: 0             타입     로스팅 포인트     지속가능성  블렌드  싱글오리진  디카페인
0        3896  [싱글오리진, 디카페인]   ['미디엄다크']        []    0      0     0
1        4603  [싱글오리진, 디카페인]   ['미디엄다크']  ['공정무역']    0      0     0
2        3460  [싱글오리진, 디카페인]     ['미디엄']   ['유기농']    0      0     0
3        4737  [싱글오리진, 디카페인]  ['라이트미디엄']        []    0      0     0
4        4701    [블렌드, 디카페인]     ['미디엄']        []    0      0     0

In [41]:
for i in range(len(decafdf)):
    for k in decafdf['타입'][i]:
        decafdf.loc[i, k] = 1
        
decafdf.head()

Unnamed: 0             타입     로스팅 포인트     지속가능성  블렌드  싱글오리진  디카페인
0        3896  [싱글오리진, 디카페인]   ['미디엄다크']        []    0      1     1
1        4603  [싱글오리진, 디카페인]   ['미디엄다크']  ['공정무역']    0      1     1
2        3460  [싱글오리진, 디카페인]     ['미디엄']   ['유기농']    0      1     1
3        4737  [싱글오리진, 디카페인]  ['라이트미디엄']        []    0      1     1
4        4701    [블렌드, 디카페인]     ['미디엄']        []    1      0     1

In [46]:
decafdf = decafdf.rename(columns={'Unnamed: 0':'id'})
decafdf.head()

id             타입     로스팅 포인트     지속가능성  블렌드  싱글오리진  디카페인
0  3896  [싱글오리진, 디카페인]   ['미디엄다크']        []    0      1     1
1  4603  [싱글오리진, 디카페인]   ['미디엄다크']  ['공정무역']    0      1     1
2  3460  [싱글오리진, 디카페인]     ['미디엄']   ['유기농']    0      1     1
3  4737  [싱글오리진, 디카페인]  ['라이트미디엄']        []    0      1     1
4  4701    [블렌드, 디카페인]     ['미디엄']        []    1      0     1

In [45]:
features[['id', '타입_블렌드', '타입_싱글오리진', '타입_디카페인']].head()

id  타입_블렌드  타입_싱글오리진  타입_디카페인
0    43       1         0        0
1  3896       0         0        1
2  3458       0         1        0
3    22       1         0        0
4   180       0         1        0

In [48]:
corrected_features = pd.merge(features, decafdf, on="id")
corrected_features.head(2)

id  바디감  신맛  단맛  쓴맛  타입_디카페인  타입_블렌드  타입_싱글오리진  지속가능성_0  지속가능성_공정무역  ...  \
0  3896    4   2   4   3        1       0         0        1           0  ...   
1  4603    3   2   3   2        1       0         0        0           1  ...   

   달콤함  고소함  향료  초콜릿             타입    로스팅 포인트     지속가능성  블렌드  싱글오리진  디카페인  
0    0    0   0    0  [싱글오리진, 디카페인]  ['미디엄다크']        []    0      1     1  
1    0    0   0    0  [싱글오리진, 디카페인]  ['미디엄다크']  ['공정무역']    0      1     1  

[2 rows x 30 columns]

In [53]:
# corrected_features.drop(columns=['타입_블렌드', '타입_싱글오리진', '타입_디카페인'], inplace=True)
corrected_features = corrected_features.rename(columns={'블렌드':'타입_블렌드', '싱글오리진':'타입_싱글오리진', '디카페인':'타입_디카페인'})
corrected_features.drop(columns=['타입', '로스팅 포인트', '지속가능성'], inplace=True)

In [54]:
corrected_features.head()

id  바디감  신맛  단맛  쓴맛  지속가능성_0  지속가능성_공정무역  지속가능성_유기농  지속가능성_직접무역  \
0  3896    4   2   4   3        1           0          0           0   
1  4603    3   2   3   2        0           1          0           0   
2  3460    3   4   5   3        0           0          1           0   
3  4737    2   4   3   1        1           0          0           0   
4  4701    3   2   4   1        1           0          0           0   

   로스팅 포인트_다크  ...  꽃  과일  허브  달콤함  고소함  향료  초콜릿  타입_블렌드  타입_싱글오리진  타입_디카페인  
0           0  ...  0   0   0    0    0   0    0       0         1        1  
1           0  ...  0   0   0    0    0   0    0       0         1        1  
2           0  ...  0   0   0    0    0   0    0       0         1        1  
3           0  ...  0   0   0    0    0   0    0       0         1        1  
4           0  ...  0   0   0    0    0   0    0       1         0        1  

[5 rows x 24 columns]

In [55]:
corrected_features = corrected_features[['id', '바디감', '신맛', '단맛', '쓴맛', '타입_디카페인', '타입_블렌드', '타입_싱글오리진',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']]
corrected_features.head(2)

id  바디감  신맛  단맛  쓴맛  타입_디카페인  타입_블렌드  타입_싱글오리진  지속가능성_0  지속가능성_공정무역  ...  \
0  3896    4   2   4   3        1       0         1        1           0  ...   
1  4603    3   2   3   2        1       0         1        0           1  ...   

   로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  달콤함  고소함  향료  초콜릿  
0               0            0              1  0   0   0    0    0   0    0  
1               0            0              1  0   0   0    0    0   0    0  

[2 rows x 24 columns]

In [56]:
# update features.csv
corrected_features.to_csv('features.csv')

In [58]:
corrected_features = corrected_features[['id', '타입_디카페인', '타입_블렌드', '타입_싱글오리진']]
corrected_features.head(2)

id  타입_디카페인  타입_블렌드  타입_싱글오리진
0  3896        1       0         1
1  4603        1       0         1

In [67]:
# need to modify mysite/data/features and connected codes for similarity score.
mysite = pd.read_csv('mysite/data/features.csv')
mysite.head()

id  body  sour  sweet  bitter  decaf  blend  타입_싱글오리진  지속가능성_0  \
0    43     4     2      4       3      0      1         0        1   
1  3896     4     2      4       3      1      0         0        1   
2  3458     3     4      5       1      0      0         1        0   
3    22     4     1      3       4      0      1         0        0   
4   180     5     3      5       1      0      0         1        1   

   지속가능성_공정무역  ...  로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  \
0           0  ...               0            0              1  0   0   0   
1           0  ...               0            0              1  0   0   0   
2           0  ...               1            0              0  0   0   0   
3           1  ...               0            0              0  0   0   0   
4           0  ...               0            0              1  0   1   0   

   달콤함  고소함  향료  초콜릿  
0    1    0   1    0  
1    0    0   0    0  
2    0    0   0    0  
3    1    1   0    1  
4    0    0   0    1  

[5 rows x 24 columns]

In [68]:
mysite.drop(columns=['decaf', 'blend', '타입_싱글오리진'], inplace=True)
mysite = pd.merge(mysite, corrected_features, on="id")

mysite = mysite.rename(columns={'타입_블렌드':'blend', '타입_싱글오리진':'single', '타입_디카페인':'decaf'})

mysite = mysite[['id', 'body', 'sour', 'sweet', 'bitter', 'decaf', 'blend', 'single',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']]

mysite.head(3)

id  body  sour  sweet  bitter  decaf  blend  single  지속가능성_0  지속가능성_공정무역  \
0  3896     4     2      4       3      1      0       1        1           0   
1  4603     3     2      3       2      1      0       1        0           1   
2  3460     3     4      5       3      1      0       1        0           0   

   ...  로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  달콤함  고소함  향료  \
0  ...               0            0              1  0   0   0    0    0   0   
1  ...               0            0              1  0   0   0    0    0   0   
2  ...               0            1              0  0   0   0    0    0   0   

   초콜릿  
0    0  
1    0  
2    0  

[3 rows x 24 columns]

In [73]:
# update mysite/data/features.csv
mysite.to_csv('mysite/data/features.csv', index=0)
mysite = pd.read_csv('mysite/data/features.csv')
mysite.head(3)  

id  body  sour  sweet  bitter  decaf  blend  single  지속가능성_0  지속가능성_공정무역  \
0  3896     4     2      4       3      1      0       1        1           0   
1  4603     3     2      3       2      1      0       1        0           1   
2  3460     3     4      5       3      1      0       1        0           0   

   ...  로스팅 포인트_라이트미디엄  로스팅 포인트_미디엄  로스팅 포인트_미디엄다크  꽃  과일  허브  달콤함  고소함  향료  \
0  ...               0            0              1  0   0   0    0    0   0   
1  ...               0            0              1  0   0   0    0    0   0   
2  ...               0            1              0  0   0   0    0    0   0   

   초콜릿  
0    0  
1    0  
2    0  

[3 rows x 24 columns]

In [31]:
import pandas as pd
decaf = pd.read_csv('decaf.csv', index_col=0)
decaf.head(3)

타입    로스팅 포인트     지속가능성
3896  ['싱글오리진', '디카페인']  ['미디엄다크']        []
4603  ['싱글오리진', '디카페인']  ['미디엄다크']  ['공정무역']
3460  ['싱글오리진', '디카페인']    ['미디엄']   ['유기농']

In [32]:
decaf['타입'] = decaf['타입'].apply(literal_eval)
decaf['타입'][3896]

['싱글오리진', '디카페인']

In [33]:
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv', index_col=0)
coffeeDB.head(2)

new_id  id   이름  로스터리ID                                              커피 소개  \
0       0   2  나탈리      60  잘 익은 딸기를 베어 문듯한 단맛과 적당한 산미의 조화를 이루는 와인의 풍미를 느낄...   
1       1   3   콜린      60  잘 익은 체리의 농후한 단맛과 고소한 아몬드 캐러멜이 입안 가득 남는 끝 맛을 느낄...   

    타입 로스팅 포인트 지속가능성                     컵 노트  바디감  ...  \
0  블렌드     미디엄  공정무역  ['딸기', '와인', '화이트 초콜릿']    2  ...   
1  블렌드   미디엄다크  공정무역  ['다크 초콜릿', '체리', '아몬드']    4  ...   

                                     나라  식품의 유형  \
0  ['파푸아뉴기니 바로이디아 콜브란', '브라질', '코스타리카']   원두 커피   
1               ['인도', '콜롬비아', '코스타리카']   원두 커피   

                                  제조원 및 소재지                       유통기한  \
0  영앤도터스, 서울 마포구 마포대로 156 공덕 프루지오시티 1층 107호                  제조일로부터 1년   
1  영앤도터스, 서울 마포구 마포대로 156 공덕 프루지오시티 1층 107호  제조일로부터 1년(권장기한 제조일로부터 1달)   

        제조일자         내용량                          보관 방법  \
0  제조일 별도 표기  200g / 1kg  직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관   
1  제조일 별도 표기  200g / 1kg  직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관   

                                  원재료 및 함량      제품문의 관련 주소 및 전화번호  price  
0  원두 커피(엘살바도르 30%, 파푸아뉴기니 40%, 코스타리카 30%)  코케 고객센터 070-4647-1868  16000  
1       원두 커피(코스타리카 30%, 콜롬비아 40%, 인도 30%)  코케 고객센터 070-4647-1868  16000  

[2 rows x 24 columns]

In [34]:
decaf['타입'][3896]

['싱글오리진', '디카페인']

In [35]:
coffeeDB[coffeeDB['id']==3896] 

new_id    id            이름  로스터리ID  \
399     399  3896  브라질 세라도 디카페인      67   

                                                 커피 소개    타입 로스팅 포인트 지속가능성  \
399  땅콩의 고소함과 코코아 파우더의 쌉싸름함 진득한 단맛과 클래식한 커피향  중강배전으...  디카페인   미디엄다크     0   

    컵 노트  바디감  ...       나라  식품의 유형               제조원 및 소재지       유통기한  \
399  NaN    4  ...  ['세라도']      커피  경기도 화성시 동탄지성로 306번길 68  제조일로부터 1년   

           제조일자       내용량                                    보관 방법   원재료 및 함량  \
399  주문 확인 후 제조  500g 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관  커피원두 100%   

      제품문의 관련 주소 및 전화번호  price  
399  운조커피 010-4535-8819  13000  

[1 rows x 24 columns]

In [36]:
for i in decaf_list:
    for k in decaf['타입'][i]:
        if k != '디카페인':
            coffeeDB.loc[coffeeDB['id'] == i, '타입'] = k
        
coffeeDB[coffeeDB['id']==3896]

new_id    id            이름  로스터리ID  \
399     399  3896  브라질 세라도 디카페인      67   

                                                 커피 소개     타입 로스팅 포인트 지속가능성  \
399  땅콩의 고소함과 코코아 파우더의 쌉싸름함 진득한 단맛과 클래식한 커피향  중강배전으...  싱글오리진   미디엄다크     0   

    컵 노트  바디감  ...       나라  식품의 유형               제조원 및 소재지       유통기한  \
399  NaN    4  ...  ['세라도']      커피  경기도 화성시 동탄지성로 306번길 68  제조일로부터 1년   

           제조일자       내용량                                    보관 방법   원재료 및 함량  \
399  주문 확인 후 제조  500g 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관  커피원두 100%   

      제품문의 관련 주소 및 전화번호  price  
399  운조커피 010-4535-8819  13000  

[1 rows x 24 columns]

In [38]:
# update mysite/data/coffeeDB2.csv
coffeeDB.to_csv('mysite/data/coffeeDB2.csv', index=0)
mysite = pd.read_csv('mysite/data/coffeeDB2.csv')
mysite.head(3)

new_id  id   이름  로스터리ID                                              커피 소개  \
0       0   2  나탈리      60  잘 익은 딸기를 베어 문듯한 단맛과 적당한 산미의 조화를 이루는 와인의 풍미를 느낄...   
1       1   3   콜린      60  잘 익은 체리의 농후한 단맛과 고소한 아몬드 캐러멜이 입안 가득 남는 끝 맛을 느낄...   
2       2   5  매뉴얼       5  아메리카노, 라떼 모두 잘 어울리는 블렌딩입니다. 부드럽고 은은하게 올라오는 산미와...   

    타입 로스팅 포인트 지속가능성                     컵 노트  바디감  ...  \
0  블렌드     미디엄  공정무역  ['딸기', '와인', '화이트 초콜릿']    2  ...   
1  블렌드   미디엄다크  공정무역  ['다크 초콜릿', '체리', '아몬드']    4  ...   
2  블렌드   미디엄다크     0                      NaN    4  ...   

                                     나라  식품의 유형  \
0  ['파푸아뉴기니 바로이디아 콜브란', '브라질', '코스타리카']   원두 커피   
1               ['인도', '콜롬비아', '코스타리카']   원두 커피   
2        ['브라질', '에티오피아', '인도', '과테말라']   볶은 커피   

                                  제조원 및 소재지                       유통기한  \
0  영앤도터스, 서울 마포구 마포대로 156 공덕 프루지오시티 1층 107호                  제조일로부터 1년   
1  영앤도터스, 서울 마포구 마포대로 156 공덕 프루지오시티 1층 107호  제조일로부터 1년(권장기한 제조일로부터 1달)   
2                                     그럼블커피                 제조일로부터 6개월   

        제조일자         내용량                          보관 방법  \
0  제조일 별도 표기  200g / 1kg  직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관   
1  제조일 별도 표기  200g / 1kg  직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관   
2  제조일 별도 표기  200g / 1kg          직사광선을 피하고 서늘한 곳에 밀폐보관   

                                  원재료 및 함량      제품문의 관련 주소 및 전화번호  price  
0  원두 커피(엘살바도르 30%, 파푸아뉴기니 40%, 코스타리카 30%)  코케 고객센터 070-4647-1868  16000  
1       원두 커피(코스타리카 30%, 콜롬비아 40%, 인도 30%)  코케 고객센터 070-4647-1868  16000  
2      커피원두 100% (브라질, 과테말라, 인도네시아, 에디오피아)  코케 고객센터 070-4647-1868  11000  

[3 rows x 24 columns]

In [40]:
mysite[mysite['id']==3896]

new_id    id            이름  로스터리ID  \
399     399  3896  브라질 세라도 디카페인      67   

                                                 커피 소개     타입 로스팅 포인트 지속가능성  \
399  땅콩의 고소함과 코코아 파우더의 쌉싸름함 진득한 단맛과 클래식한 커피향  중강배전으...  싱글오리진   미디엄다크     0   

    컵 노트  바디감  ...       나라  식품의 유형               제조원 및 소재지       유통기한  \
399  NaN    4  ...  ['세라도']      커피  경기도 화성시 동탄지성로 306번길 68  제조일로부터 1년   

           제조일자       내용량                                    보관 방법   원재료 및 함량  \
399  주문 확인 후 제조  500g 1kg  직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관  커피원두 100%   

      제품문의 관련 주소 및 전화번호  price  
399  운조커피 010-4535-8819  13000  

[1 rows x 24 columns]

***
add cupnote category to coffeeDB
***

In [41]:
category = pd.read_csv('category.csv')
category.head(3)

id                 컵 노트  꽃  과일  허브  달콤함  고소함  향료  초콜릿
0    43  ['토피', '맥아', '흑설탕']  0   0   0    1    0   1    0
1  3896             ['null']  0   0   0    0    0   0    0
2  3458             ['null']  0   0   0    0    0   0    0

In [44]:
category = category[['id', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']]
category.head(3)

id  꽃  과일  허브  달콤함  고소함  향료  초콜릿
0    43  0   0   0    1    0   1    0
1  3896  0   0   0    0    0   0    0
2  3458  0   0   0    0    0   0    0

In [45]:
coffeeDB = pd.merge(mysite, category, on="id")
coffeeDB.head(2)

new_id  id   이름  로스터리ID                                              커피 소개  \
0       0   2  나탈리      60  잘 익은 딸기를 베어 문듯한 단맛과 적당한 산미의 조화를 이루는 와인의 풍미를 느낄...   
1       1   3   콜린      60  잘 익은 체리의 농후한 단맛과 고소한 아몬드 캐러멜이 입안 가득 남는 끝 맛을 느낄...   

    타입 로스팅 포인트 지속가능성                     컵 노트  바디감  ...  \
0  블렌드     미디엄  공정무역  ['딸기', '와인', '화이트 초콜릿']    2  ...   
1  블렌드   미디엄다크  공정무역  ['다크 초콜릿', '체리', '아몬드']    4  ...   

                                  원재료 및 함량      제품문의 관련 주소 및 전화번호  price  꽃  \
0  원두 커피(엘살바도르 30%, 파푸아뉴기니 40%, 코스타리카 30%)  코케 고객센터 070-4647-1868  16000  0   
1       원두 커피(코스타리카 30%, 콜롬비아 40%, 인도 30%)  코케 고객센터 070-4647-1868  16000  0   

  과일 허브 달콤함 고소함 향료 초콜릿  
0  1  0   0   0  0   1  
1  1  0   0   1  0   1  

[2 rows x 31 columns]

In [47]:
# update mysite/data/coffeeDB2.csv
coffeeDB.to_csv('mysite/data/coffeeDB2.csv', index=0)
mysite = pd.read_csv('mysite/data/coffeeDB2.csv')
mysite.head(3)

new_id  id   이름  로스터리ID                                              커피 소개  \
0       0   2  나탈리      60  잘 익은 딸기를 베어 문듯한 단맛과 적당한 산미의 조화를 이루는 와인의 풍미를 느낄...   
1       1   3   콜린      60  잘 익은 체리의 농후한 단맛과 고소한 아몬드 캐러멜이 입안 가득 남는 끝 맛을 느낄...   
2       2   5  매뉴얼       5  아메리카노, 라떼 모두 잘 어울리는 블렌딩입니다. 부드럽고 은은하게 올라오는 산미와...   

    타입 로스팅 포인트 지속가능성                     컵 노트  바디감  ...  \
0  블렌드     미디엄  공정무역  ['딸기', '와인', '화이트 초콜릿']    2  ...   
1  블렌드   미디엄다크  공정무역  ['다크 초콜릿', '체리', '아몬드']    4  ...   
2  블렌드   미디엄다크     0                      NaN    4  ...   

                                  원재료 및 함량      제품문의 관련 주소 및 전화번호  price  꽃  \
0  원두 커피(엘살바도르 30%, 파푸아뉴기니 40%, 코스타리카 30%)  코케 고객센터 070-4647-1868  16000  0   
1       원두 커피(코스타리카 30%, 콜롬비아 40%, 인도 30%)  코케 고객센터 070-4647-1868  16000  0   
2      커피원두 100% (브라질, 과테말라, 인도네시아, 에디오피아)  코케 고객센터 070-4647-1868  11000  0   

  과일 허브 달콤함 고소함 향료 초콜릿  
0  1  0   0   0  0   1  
1  1  0   0   1  0   1  
2  0  0   0   0  0   0  

[3 rows x 31 columns]

***
식품의 유형
***

In [48]:
df = pd.read_csv('wondoofin.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 559 non-null    int64 
 1   이름                 559 non-null    object
 2   로스터리               559 non-null    object
 3   타입                 559 non-null    object
 4   로스팅 포인트            559 non-null    object
 5   지속가능성              559 non-null    object
 6   컵 노트               559 non-null    object
 7   바디감                559 non-null    int64 
 8   신맛                 559 non-null    int64 
 9   단맛                 559 non-null    int64 
 10  쓴맛                 559 non-null    int64 
 11  커피 소개              559 non-null    object
 12  나라                 559 non-null    object
 13  로스터리 소개            559 non-null    object
 14  로스터리 주소            559 non-null    object
 15  이름.1               559 non-null    object
 16  식품의 유형             554 non-null    object
 1

In [49]:
df['식품의 유형'].describe()

count     554
unique     13
top        커피
freq      426
Name: 식품의 유형, dtype: object

In [50]:
df['식품의 유형'].unique()

array(['커피(볶은커피)', '커피', '볶은커피', '원두커피', '원두', '볶은 커피', '디카페인 커피',
       '커피 원두', '커피원두', nan, '커피원', '원두 커피', '가공 원두', '커피 원두 100%'],
      dtype=object)

In [51]:
df['식품의 유형'] = '커피 원두'

In [53]:
# update wondoofin.csv
df.to_csv('wondoofin.csv', index=0)
df = pd.read_csv('wondoofin.csv')
df['식품의 유형'].describe()

count       559
unique        1
top       커피 원두
freq        559
Name: 식품의 유형, dtype: object

In [57]:
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv')
coffeeDB['식품의 유형'].describe()

count     554
unique     13
top        커피
freq      426
Name: 식품의 유형, dtype: object

In [58]:
coffeeDB['식품의 유형'] = '커피 원두'
coffeeDB.head(2)

new_id  id   이름  로스터리ID                                              커피 소개  \
0       0   2  나탈리      60  잘 익은 딸기를 베어 문듯한 단맛과 적당한 산미의 조화를 이루는 와인의 풍미를 느낄...   
1       1   3   콜린      60  잘 익은 체리의 농후한 단맛과 고소한 아몬드 캐러멜이 입안 가득 남는 끝 맛을 느낄...   

    타입 로스팅 포인트 지속가능성                     컵 노트  바디감  ...  \
0  블렌드     미디엄  공정무역  ['딸기', '와인', '화이트 초콜릿']    2  ...   
1  블렌드   미디엄다크  공정무역  ['다크 초콜릿', '체리', '아몬드']    4  ...   

                                  원재료 및 함량      제품문의 관련 주소 및 전화번호  price  꽃  \
0  원두 커피(엘살바도르 30%, 파푸아뉴기니 40%, 코스타리카 30%)  코케 고객센터 070-4647-1868  16000  0   
1       원두 커피(코스타리카 30%, 콜롬비아 40%, 인도 30%)  코케 고객센터 070-4647-1868  16000  0   

  과일 허브 달콤함 고소함 향료 초콜릿  
0  1  0   0   0  0   1  
1  1  0   0   1  0   1  

[2 rows x 31 columns]

In [59]:
# update mysite/data/coffeeDB2.csv
coffeeDB.to_csv('mysite/data/coffeeDB2.csv', index=0)
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv')
coffeeDB['식품의 유형'].describe()

count       559
unique        1
top       커피 원두
freq        559
Name: 식품의 유형, dtype: object

***
유통기한
***
- count : 557/559
- 같은 말인데 다르게 쓰여 있는 value들 합치기

In [60]:
df = pd.read_csv('wondoofin.csv')
df['유통기한'].describe()

count           557
unique           23
top       제조일로부터 1년
freq            452
Name: 유통기한, dtype: object

In [61]:
df['유통기한'].unique()

array(['제조일로부터 12개월', '제조일로부터 1년', '제품 별도 표기', '제조일로부터 6개월',
       '제조일로부터 2달 이내', '제조일로 부터 1년', '제조일로부터 3개월 이내(실온 상태)', nan,
       '제조일로부터1년', '제조일로부터 1개월', '제조일로부터 3개월까지(실온상태)', '제조일로부터 3개월 이내',
       '제조일로부터24개월', '제조일자로부터 1년', '제조일부터 1년 이내', '제조일로 2개월',
       '제조일로부터 2개월', '제조일로부터 12개월 이내', '제조일로부터 24개월', '제조일로부터 3개월',
       '제조일로부터 10개월', '제조일로부터 1년(권장기한 제조일로부터 1달)',
       '제조일로부터 1년(향미 보존 위해 3주 이내에 소비 권장)', '제조일로부터 6개월 이내 권장'],
      dtype=object)

'제조일로부터 1년' : '제조일로부터 12개월', '제조일로 부터 1년', '제조일로부터1년', '제조일자로부터 1년', '제조일부터 1년 이내', '제조일로부터 12개월 이내', '제조일로부터 1년(권장기한 제조일로부터 1달)', '제조일로부터 1년(향미 보존 위해 3주 이내에 소비 권장)'

'제품 별도 표기' : nan

'제조일로부터 6개월' : '제조일로부터 6개월 이내 권장'

'제조일로부터 2개월' : '제조일로부터 2달 이내', '제조일로 2개월'

'제조일로부터 3개월' : '제조일로부터 3개월 이내(실온 상태)', '제조일로부터 3개월까지(실온상태)', '제조일로부터 3개월 이내'

'제조일로부터 1개월'  

'제조일로부터 24개월' : '제조일로부터24개월'

'제조일로부터 10개월'

In [69]:
df['유통기한'] = df['유통기한'].map(lambda x: 
    '제조일로부터 1년' if x in [
        '제조일로부터 12개월', '제조일로 부터 1년', '제조일로부터1년', '제조일자로부터 1년',
        '제조일부터 1년 이내', '제조일로부터 12개월 이내', '제조일로부터 1년(권장기한 제조일로부터 1달)',
        '제조일로부터 1년(향미 보존 위해 3주 이내에 소비 권장)'
    ] else (
    '제품 별도 표기' if pd.isna(x) or x == '제품 별도 표기' else (
    '제조일로부터 2개월' if x == '제조일로부터 2달 이내' or x == '제조일로 2개월' else (
    '제조일로부터 6개월' if x == '제조일로부터 6개월 이내 권장' else (
    '제조일로부터 24개월' if x == '제조일로부터24개월' else (
    '제조일로부터 3개월' if x in ['제조일로부터 3개월 이내(실온 상태)', '제조일로부터 3개월까지(실온상태)', '제조일로부터 3개월 이내'] else x
    )))))
)
df['유통기한'].unique()

array(['제조일로부터 1년', '제품 별도 표기', '제조일로부터 6개월', '제조일로부터 2개월', '제조일로부터 3개월',
       '제조일로부터 1개월', '제조일로부터 24개월', '제조일로부터 10개월'], dtype=object)

In [70]:
# update wondoofin.csv
df.to_csv('wondoofin.csv', index=0)
df = pd.read_csv('wondoofin.csv')
df['유통기한'].unique()

array(['제조일로부터 1년', '제품 별도 표기', '제조일로부터 6개월', '제조일로부터 2개월', '제조일로부터 3개월',
       '제조일로부터 1개월', '제조일로부터 24개월', '제조일로부터 10개월'], dtype=object)

In [79]:
coffee = pd.read_csv('mysite/data/coffeeDB2.csv')
coffee.columns

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [80]:
coffeeDB = pd.merge(coffee, df[['id', '유통기한']], on='id')
coffeeDB.columns

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한_x', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿', '유통기한_y'],
      dtype='object')

In [81]:
coffeeDB.drop(columns=['유통기한_x'], inplace=True)
coffeeDB = coffeeDB.rename(columns={'유통기한_y': '유통기한'})

column_order = ['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']

coffeeDB = coffeeDB[column_order]

print(coffeeDB.columns)
print(coffeeDB['유통기한'].unique())

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')
['제조일로부터 1년' '제조일로부터 6개월' '제조일로부터 2개월' '제조일로부터 1개월' '제조일로부터 3개월'
 '제조일로부터 10개월' '제조일로부터 24개월' '제품 별도 표기']


In [82]:
# update mysite/data/coffeeDB2.csv
coffeeDB.to_csv('mysite/data/coffeeDB2.csv', index=0)
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv')
coffeeDB['유통기한'].unique()

array(['제조일로부터 1년', '제조일로부터 6개월', '제조일로부터 2개월', '제조일로부터 1개월',
       '제조일로부터 3개월', '제조일로부터 10개월', '제조일로부터 24개월', '제품 별도 표기'],
      dtype=object)

***
보관 방법
***

In [83]:
df = pd.read_csv('wondoofin.csv')
df['보관 방법'].describe()

count                                         545
unique                                         36
top       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관
freq                                          358
Name: 보관 방법, dtype: object

In [84]:
df['보관 방법'].unique()

array(['직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.',
       '직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관',
       '직사광선을 피하고 서늘한 곳에 보관하세요.', '건냉한 곳에 밀폐 보관', '직사광선을 피하고 밀봉하여 실온보관',
       '직사광선을 피하고 서늘하고 건조한 곳', '실온보관', nan, '건냉한 곳에 밀폐보관', '없음',
       '서늘한 곳에 밀폐보관', '서늘한 곳에 보관', '밀봉 후 직사광선을 피하여 실온 보관', '해당없음',
       '직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관',
       '직사광선과 습기를 피하여 서늘한 곳에 실온 보관 하며, 개봉한 후에는 빨리 드십시오.',
       '직사광선을 피하고, 온도 습도가 낮으며 통풍이 잘되는 곳에 밀폐 보관해 주십시오.',
       '실온에서 직사광선을 피해 서늘한 곳에 보관하세요.', '습기 및 직사광선을 피하고 서늘한 장소에 보관하십시오.',
       '건조하고 서늘한 곳에 보관', '직사광선을 피하고 밀봉하여 실온', '직사광선을 피하고 서늘한 곳에 보관하십시오.',
       '직사광선을 피하고 서늘한 곳에 밀폐보관', '-',
       '직사광선을 피해 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하십시오.',
       '직사광선을 피하고 서늘하고 건조한 곳에 보관하십시오. 개봉 후 빨리 드시기 바랍니다.',
       '햇빛이 들지 않는 상온에 밀폐 보관',
       '개봉 후 서늘한 곳에 밀폐하여 보관하시고, 되도록 공기와의 접촉을 피해야합니다.', '상온 보관', '실온 보관',
       '후면에 표기', '공기가 들어가지 않도록 밀봉 또는 밀폐 용기에 담아 서늘한 그늘에 보관하십시오.',
       '직사광선 피하고 서늘한 곳', '서늘한 곳에 상온보관', '직사광선을 피하고 서늘한 곳에 보관하십시오',
       '직

In [85]:
df['보관 방법'] = df['보관 방법'].map(lambda x: 
    '제조일로부터 1년' if pd.isna(x) or x in [
        '없음', '해당없음', '-'
    ] else x
    )
df['보관 방법'].unique()

array(['직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하여 주십시오.',
       '직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관',
       '직사광선을 피하고 서늘한 곳에 보관하세요.', '건냉한 곳에 밀폐 보관', '직사광선을 피하고 밀봉하여 실온보관',
       '직사광선을 피하고 서늘하고 건조한 곳', '실온보관', '제조일로부터 1년', '건냉한 곳에 밀폐보관',
       '서늘한 곳에 밀폐보관', '서늘한 곳에 보관', '밀봉 후 직사광선을 피하여 실온 보관',
       '직사광선을 피하고 밀봉하여 건조하고 서늘한 곳에 보관',
       '직사광선과 습기를 피하여 서늘한 곳에 실온 보관 하며, 개봉한 후에는 빨리 드십시오.',
       '직사광선을 피하고, 온도 습도가 낮으며 통풍이 잘되는 곳에 밀폐 보관해 주십시오.',
       '실온에서 직사광선을 피해 서늘한 곳에 보관하세요.', '습기 및 직사광선을 피하고 서늘한 장소에 보관하십시오.',
       '건조하고 서늘한 곳에 보관', '직사광선을 피하고 밀봉하여 실온', '직사광선을 피하고 서늘한 곳에 보관하십시오.',
       '직사광선을 피하고 서늘한 곳에 밀폐보관',
       '직사광선을 피해 온도, 습도가 낮으며 통풍이 잘 되는 곳에 보관하십시오.',
       '직사광선을 피하고 서늘하고 건조한 곳에 보관하십시오. 개봉 후 빨리 드시기 바랍니다.',
       '햇빛이 들지 않는 상온에 밀폐 보관',
       '개봉 후 서늘한 곳에 밀폐하여 보관하시고, 되도록 공기와의 접촉을 피해야합니다.', '상온 보관', '실온 보관',
       '후면에 표기', '공기가 들어가지 않도록 밀봉 또는 밀폐 용기에 담아 서늘한 그늘에 보관하십시오.',
       '직사광선 피하고 서늘한 곳', '서늘한 곳에 상온보관', '직사광선을 피하고 서늘한 곳에 보관하십시오',
       '직사광선을 피하고, 건

In [86]:
df['보관 방법'].describe()

count                                         559
unique                                         34
top       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관
freq                                          358
Name: 보관 방법, dtype: object

In [87]:
# update wondoofin.csv
df.to_csv('wondoofin.csv', index=0)
df = pd.read_csv('wondoofin.csv')
df['보관 방법'].describe()

count                                         559
unique                                         34
top       직사광선을 피하고 온도, 습도가 낮으며 통풍이 잘 되는 곳에 밀폐 보관
freq                                          358
Name: 보관 방법, dtype: object

In [98]:
coffee = pd.read_csv('mysite/data/coffeeDB2.csv')
coffee.columns

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [99]:
coffeeDB = pd.merge(coffee, df[['id', '보관 방법']], on='id')
coffeeDB.columns

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법_x', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿', '보관 방법_y'],
      dtype='object')

In [100]:
print(len(coffeeDB['보관 방법_x'].unique()))
print(len(coffeeDB['보관 방법_y'].unique()))

37
34


In [101]:
coffeeDB.drop(columns=['보관 방법_x'], inplace=True)
coffeeDB = coffeeDB.rename(columns={'보관 방법_y': '보관 방법'})

column_order = ['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']

coffeeDB = coffeeDB[column_order]

print(coffeeDB.columns)
print(len(coffeeDB['보관 방법'].unique()))

Index(['new_id', 'id', '이름', '로스터리ID', '커피 소개', '타입', '로스팅 포인트', '지속가능성',
       '컵 노트', '바디감', '신맛', '단맛', '쓴맛', '카페인', '나라', '식품의 유형', '제조원 및 소재지',
       '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량', '제품문의 관련 주소 및 전화번호',
       'price', '꽃', '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')
34


In [102]:
# update mysite/data/coffeeDB2.csv
coffeeDB.to_csv('mysite/data/coffeeDB2.csv', index=0)
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv')
print(len(coffeeDB['보관 방법'].unique()))

34


***
확인
***

In [157]:
df = pd.read_csv('wondoofin.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 559 non-null    int64 
 1   이름                 559 non-null    object
 2   로스터리               559 non-null    object
 3   타입                 559 non-null    object
 4   로스팅 포인트            559 non-null    object
 5   지속가능성              559 non-null    object
 6   컵 노트               559 non-null    object
 7   바디감                559 non-null    int64 
 8   신맛                 559 non-null    int64 
 9   단맛                 559 non-null    int64 
 10  쓴맛                 559 non-null    int64 
 11  커피 소개              559 non-null    object
 12  나라                 559 non-null    object
 13  로스터리 소개            559 non-null    object
 14  로스터리 주소            559 non-null    object
 15  이름.1               559 non-null    object
 16  식품의 유형             559 non-null    object
 1

In [158]:
df.columns

Index(['id', '이름', '로스터리', '타입', '로스팅 포인트', '지속가능성', '컵 노트', '바디감', '신맛', '단맛',
       '쓴맛', '커피 소개', '나라', '로스터리 소개', '로스터리 주소', '이름.1', '식품의 유형',
       '제조원 및 소재지', '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량',
       '제품문의 관련 주소 및 전화번호', '카페인', '로스터리ID', 'price'],
      dtype='object')

In [159]:
df['지속가능성'].unique()

array(['0', '유기농', '공정무역', '직접무역'], dtype=object)

In [160]:
import numpy as np
df['컵 노트'].replace("['null']", np.nan, inplace=True)
df['컵 노트'][1]

nan

In [162]:
flavour = ['바디감', '신맛', '단맛', '쓴맛']
for i in flavour:
    print(df[i].unique())

[4 3 5 2 1]
[2 4 1 3 5]
[4 5 3 2]
[3 1 4 2 5]


'커피 소개', '나라', '로스터리 소개', '로스터리 주소', '이름.1', '식품의 유형',
       '제조원 및 소재지', '유통기한', '제조일자', '내용량', '보관 방법', '원재료 및 함량',
       '제품문의 관련 주소 및 전화번호', '카페인', '로스터리ID', 'price'

***
디카페인 수정할 때 실수로 feature.csv를 디카페인인 원두들만의 정보만 들어있는 dataframe으로 덮어 써버려서 되돌리고 최신 정보로 수정
***

In [145]:
df = pd.read_csv('mysite/data/features.csv')
df.columns

Index(['id', 'body', 'sour', 'sweet', 'bitter', 'decaf', 'blend', 'single',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')

In [146]:
coffeeDB = pd.read_csv('mysite/data/coffeeDB2.csv')
coffeeDB['타입'].unique()

array(['블렌드', '싱글오리진'], dtype=object)

In [147]:
features = pd.get_dummies(coffeeDB, columns=['타입'], dtype=int)

In [148]:
features = features[['id', '로스팅 포인트', '지속가능성','카페인', '바디감', '신맛', '단맛', '쓴맛', '타입_블렌드', '타입_싱글오리진', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']]
features = pd.get_dummies(features, columns=['지속가능성', '로스팅 포인트'], dtype=int)

In [149]:
print(features.columns)
print(len(features.columns))

Index(['id', '카페인', '바디감', '신맛', '단맛', '쓴맛', '타입_블렌드', '타입_싱글오리진', '꽃', '과일',
       '허브', '달콤함', '고소함', '향료', '초콜릿', '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농',
       '지속가능성_직접무역', '로스팅 포인트_다크', '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄',
       '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크'],
      dtype='object')
24


In [150]:
print(df.columns)
print(len(df.columns))

Index(['id', 'body', 'sour', 'sweet', 'bitter', 'decaf', 'blend', 'single',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿'],
      dtype='object')
24


In [151]:
column_mapping = {'바디감' : 'body', '신맛' : 'sour', '단맛' : 'sweet', '쓴맛' : 'bitter', '카페인' : 'caf', '타입_블렌드' : 'blend', '타입_싱글오리진':'single'}
features.rename(columns=column_mapping, inplace=True)

column_order = ['id', 'body', 'sour', 'sweet', 'bitter', 'caf', 'blend', 'single',
       '지속가능성_0', '지속가능성_공정무역', '지속가능성_유기농', '지속가능성_직접무역', '로스팅 포인트_다크',
       '로스팅 포인트_라이트', '로스팅 포인트_라이트미디엄', '로스팅 포인트_미디엄', '로스팅 포인트_미디엄다크', '꽃',
       '과일', '허브', '달콤함', '고소함', '향료', '초콜릿']
features = features[column_order]

print(features.columns == df.columns)

[ True  True  True  True  True False  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True]


In [154]:
print(features.shape)
print(df.shape)

(559, 24)
(93, 24)


In [155]:
features.to_csv('mysite/data/features.csv', index=0)
featuresDB = pd.read_csv('mysite/data/features.csv')
featuresDB.shape

(559, 24)